src/modules/mc_att_control/AttitudeControl/AttitudeControlTest.cpp

```diff
@@ -33,12 +33,95 @@
 
 #include <gtest/gtest.h>
 #include <AttitudeControl.hpp>
+#include <mathlib/math/Functions.hpp>
 
 using namespace matrix;
 
 TEST(AttitudeControlTest, AllZeroCase)
 {
 	AttitudeControl attitude_control;
-	matrix::Vector3f rate_setpoint = attitude_control.update(Quatf(), Quatf(), 0.f);
+	Vector3f rate_setpoint = attitude_control.update(Quatf(), Quatf(), 0.f);
 	EXPECT_EQ(rate_setpoint, Vector3f());
 }
+
+class AttitudeControlConvergenceTest : public ::testing::Test
+{
+public:
+	AttitudeControlConvergenceTest()
+	{
+		_attitude_control.setProportionalGain(Vector3f(.5f, .6f, .3f));
+		_attitude_control.setRateLimit(Vector3f(100, 100, 100));
+	}
+
+	void checkConvergence()
+	{
+		int i; // need function scope to check how many steps
+		Vector3f rate_setpoint(1000, 1000, 1000);
+
+		for (i = 100; i > 0; i--) {
+			// run attitude control to get rate setpoints
+			const Vector3f rate_setpoint_new = _attitude_control.update(_quat_state, _quat_goal, 0.f);
+			// rotate the simulated state quaternion according to the rate setpoint
+			_quat_state = _quat_state * Quatf(AxisAnglef(rate_setpoint_new));
+
+			// expect the error and hence also the output to get smaller with each iteration
+			if (rate_setpoint_new.norm() >= rate_setpoint.norm()) {
+				break;
+			}
+
+			rate_setpoint = rate_setpoint_new;
+		}
+
+		// we need to have reached the goal attitude
+		if (!(antipodal(_quat_state) == (antipodal(_quat_goal)))) {
+			antipodal(_quat_state).print();
+			antipodal(_quat_goal).print();
+		}
+
+		EXPECT_TRUE(antipodal(_quat_state) == antipodal(_quat_goal));
+		// it shouldn't have taken longer than an iteration timeout to converge
+		EXPECT_GT(i, 0);
+	}
+
+	Quatf antipodal(const Quatf q)
+	{
+		for (int i = 0; i < 1; i++) {
```

Comment:
Do you really want to check the real part only?

Reply:
No, this was a temporary change that I accidentally commited 😱 Good catch!
And it was still passing the tests because of a second bug in this function: It returned a NAN quaternion if none of the above cases apply which should in practise never happen because the quaternion needs to have unit length and hence is longer than 0. But it actually happened because I didn't check all the cases in the loop and since the function is applied to both sides the test passed.

Reply:
Both mentioned bugs fixed in https://github.com/PX4/Firmware/pull/13102/commits/24d6e4e03b9eebb5632cdd15a82d93500879e216
I checked that now every single loop iteration is necessary to pass the tests.